eclass/haskell-cabal.eclass

```diff
@@ -445,7 +445,7 @@ cabal-build() {
 }
 
 cabal-copy() {
-	set -- copy --destdir="${D}" "$@"
+	set -- copy "$@" --destdir="${D}"
```

Comment:
Could you please explain the motivation for this change, and what it does? Just so that I can feel more at ease that this is not going to break anything.

Reply:
No problem. My motivation stems from the `ghcide` package which has some extra executables which I have turned off by default. You can see the USE flags here:

https://github.com/gentoo-haskell/gentoo-haskell/pull/1236/commits/af9c3c26bd0d43ceb4500937ef07e41066152067#diff-339038abe81bfac3da07a4a0a1198e4b72942844c1a7268f062ff703adf19e49R22

In the [ghcide cabal file]() you can see that the tests require the `ghcide` and `ghcide-test-preprocessor` executables:

```
test-suite ghcide-tests
    type: exitcode-stdio-1.0
    default-language: Haskell2010
    build-tool-depends:
        ghcide:ghcide,
        ghcide:ghcide-test-preprocessor,
        implicit-hie:gen-hie
```

Originally, I solved this by using `REQUIRED_USE="test? (executable test-exe)`, but this wasn't ideal because someone might want to enable the tests but not want those executables installed on the system.

My first attempt to deal with this little thorn was to have configure run with `--flag=executable` (for instance) if `USE=test` _or_ `USE=executable` were enabled, then check to see if `USE=executable` was enabled during the install phase and manually `rm` the executable if it wasn't.

Well, it turns out the `./setup copy` command can take a list of "components" such as `lib:ghcide` or `ghcide:exe:ghcide-test-preprocessor` which controls what components get installed:

```
# ./setup copy --help
Copy the files of all/specific components to install locations.

Usage: setup copy [FLAGS]
   or: setup copy COMPONENTS [FLAGS]

Components encompass executables and libraries. Does not call register, and
allows a prefix at install time. Without the --destdir flag, configure
determines location.

Flags for copy:
 -h --help         Show this help text
 -v --verbose[=n]  Control verbosity (n is 0--3, default verbosity level is 1)
    --builddir=DIR The directory where Cabal puts generated build files
                   (default dist)
    --destdir=DIR  directory to copy files to, prepended to installation
                   directories

Examples:
  setup copy               All the components in the package
  setup copy foo           A component (i.e. lib, exe, test suite)
```

If no components are passed to this command, the default behavior is to install all of the libraries and executables. As you can see, arguments passed to `cabal_src_install()` were originally dropped, but now get passed to `cabal-copy()`:

https://github.com/gentoo-haskell/gentoo-haskell/pull/1236/files/d3d84bb3ad7c1b76054bef200f797d955cc46293#diff-62462e7839e1185bef5ca4caeb99d317e8b97d071cbcc512ec7d2b70e65d7a50L597-L612

This allows you to pass arguments to `haskell-cabal_src_install()` and they will make their way to `./setup copy`. However, if no arguments are passed, the behavior will be unchanged from before.

Reply:
It turned out that I couldn't use the tests for `ghcide` anyway, because they eventually start making network calls and it would be a big effort to try to make them work in Gentoo. But, I thought the change would be useful in the future and would not effect current ebuilds.

Reply:
Oh, I am using this change for `hiedb-0.4.1.0.ebuild`:

https://github.com/gentoo-haskell/gentoo-haskell/blob/d3d84bb3ad7c1b76054bef200f797d955cc46293/dev-haskell/hiedb/hiedb-0.4.1.0.ebuild#L44-L52

https://github.com/gentoo-haskell/gentoo-haskell/blob/d3d84bb3ad7c1b76054bef200f797d955cc46293/dev-haskell/hiedb/hiedb-0.4.1.0.ebuild#L54-L62

Reply:
Thanks for the very thorough explainer! I feel comfortable with it, from what you've told me. We can test it in here for a few months and then upstream it to `::gentoo` when we show that no bugs arise from the change.

Other than that, I feel like you know these packages much better than me, so if you're happy to merge these in, I'm happy.